libpromises/evalfunction.c

```diff
@@ -5339,15 +5322,95 @@ static FnCallResult FnCallDatatype(EvalContext *ctx, ARG_UNUSED const Policy *po
             subtype_str = "null";
             break;
         default:
-            Log(LOG_LEVEL_ERR,
-                "Function %s failed to get subtype of type data", fp->name);
-            return FnFailure();
+            return NULL;
         }
 
-        return FnReturnF("%s %s", type_str, subtype_str);
+        return StringConcatenate(3, type_str, " ", subtype_str);
     }
+    return StringConcatenate(2, "policy ", type_str);
+}
+
+static FnCallResult FnCallDatatype(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
 
-    return FnReturnF("policy %s", type_str);
+    if (finalargs == NULL)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Function %s requires variable identifier as first argument",
+            fp->name);
+        return FnFailure();
+    }
+    const char* const var_name = RlistScalarValue(finalargs);
```

Comment:
We put the asterisk close to the const, to emphasize that it's the pointer which is const, not what it's pointing to.

```suggestion
    const char *const var_name = RlistScalarValue(finalargs);
```